modules/widgets/instagram.php

```diff
@@ -0,0 +1,497 @@
+
+		$data = $this->get_data( $instance );
+		// TODO: Revisit the error handling. I think we should be using WP_Error here and
+		// Jetpack::Client is the legacy check
+		if ( 'ERROR' === $data || 'ERROR' === $instance['is_legacy_token'] ) {
+			echo '<p>' . __( 'Instagram is currently experiencing connectivity issues, please try again later to connect.', 'jetpack' ) . '</p>';
+			return;
+		}
+		echo '<p>' . sprintf( __( '<strong>Connected Instagram Account</strong><br /> <a href="%1$s">%2$s</a> | <a href="%3$s">remove</a>', 'jetpack' ), esc_url( 'http://instagram.com/' . $data['external_name'] ), esc_html( $data['external_name'] ), esc_url( $remove_token_id_url ) ) . '</p>';
+
+		// Title
+		echo '<p><label><strong>' . __( 'Widget Title', 'jetpack' ) . '</strong> <input type="text" id="' . esc_attr( $this->get_field_id( 'title' ) ) . '" name="' . esc_attr( $this->get_field_name( 'title' ) ) . '" value="' . esc_attr( $instance['title'] ) . '" class="widefat" /></label></p>';
+
+		// Number of images to show
+		echo '<p><label>';
+		echo '<strong>' . __( 'Images', 'jetpack' ) . '</strong><br />';
+		echo __( 'Number to display:', 'jetpack' ) . ' ';
+		echo '<select name="' . esc_attr( $this->get_field_name( 'count' ) ) . '">';
+		for ( $i = 1; $i <= $this->valid_options['max_count']; $i++ ) {
+			echo '<option value="' . esc_attr( $i ) . '"' . selected( $i, $instance['count'], false ) . '>' . $i . '</option>';
+		}
+		echo '</select>';
+		echo '</label></p>';
+
+		// Columns
+		echo '<p><label>';
+		echo '<strong>' . __( 'Layout', 'jetpack' ) . '</strong><br />';
+		echo __( 'Number of columns:', 'jetpack' ) . ' ';
+		echo '<select name="' . esc_attr( $this->get_field_name( 'columns' ) ) . '">';
+		for ( $i = 1; $i <= $this->valid_options['max_columns']; $i++ ) {
+			echo '<option value="' . esc_attr( $i ) . '"' . selected( $i, $instance['columns'], false ) . '>' . $i . '</option>';
+		}
+		echo '</select>';
+		echo '</label></p>';
+
+		echo '<p><small>' . sprintf( __( 'New images may take up to %d minutes to show up on your site.', 'jetpack' ), 15 ) . '</small></p>';
+	}
+
+	/**
+	 * Validates and sanitizes the user-supplied widget options.
+	 *
+	 * @param array $new_instance The user-supplied values.
+	 * @param array $old_instance The existing widget options.
+	 * @return array A validated and sanitized version of $new_instance.
+	 */
+	public function update( $new_instance, $old_instance ) {
+		$instance = $this->defaults;
+
+		$instance['token_id'] = $old_instance['token_id'];
+
+		$instance['title'] = strip_tags( $new_instance['title'] );
+
+		$instance['columns'] = max( 1, min( $this->valid_options['max_columns'], (int) $new_instance['columns'] ) );
+
+		$instance['count'] = max( 1, min( $this->valid_options['max_count'], (int) $new_instance['count'] ) );
+
+		return $instance;
+	}
+}
+
+add_action( 'widgets_init', function() {
+	register_widget( 'Jetpack_Instagram_Widget' );
```

Comment:
Since this widget relies on a connection to WordPress.com, I'd recommend adding a `Jetpack::is_active()` check here.